### ISSUE TITLE
Improve group-switching AH models: contribution trajectory and switch recall

### ISSUE THREAD

**Issue by cemrtkn** [LBrinkmann/algorithmic-institutions#78]
## Context

After retraining on the corrected dataset (#76), a simulation run against the updated pilot data shows the current model stack does not reproduce the pilot's group dynamics. This issue tracks the modelling/training work needed to close the gap.

## Observed gaps (pilot vs simulation with #76 models)

**1. Contribution trajectory is flat.**
Pilot contribution rises from ~8 → ~10 across 24 rounds; simulated contribution stays flat at ~8. Downstream: common_good stuck at ~11 (pilot ~22), payoff at ~21 (pilot ~29).

![comparison_pilot_vs_sim](https://github.com/center-for-humans-and-machines/algorithmic-institutions/blob/78-improve-ah-models-contribution-switch/plots/simulation/ah_group_switching_punishment/comparison_pilot_vs_sim.jpg?raw=true)

**2. Switch predictor under-predicts by ~6–9×.**
At each of the 5 switch decision rounds (every 4 rounds), pilot shows ~2.5–3.5 switches; sim shows ~0.4. The binary classifier is collapsing to the majority "no-switch" class.

![switch_count_per_round](https://github.com/center-for-humans-and-machines/algorithmic-institutions/blob/78-improve-ah-models-contribution-switch/plots/simulation/ah_group_switching_punishment/switch_count_per_round.jpg?raw=true)

![group_size_evolution](https://github.com/center-for-humans-and-machines/algorithmic-institutions/blob/78-improve-ah-models-contribution-switch/plots/simulation/ah_group_switching_punishment/group_size_evolution_global.jpg?raw=true)

## Contribution model investigation

- Grid search over architecture, hidden_size, lr, and weight_decay
- Compare training dynamics across 3 runs: legacy data + legacy params, new data + legacy params, new data + optimized params
- Data analysis: compare dataset statistics, feature importance between legacy and new data
- Preprocessing review of `scripts/data_creation/group_switching_preprocess.py`

Details in comments: [grid search & training dynamics](https://github.com/center-for-humans-and-machines/algorithmic-institutions/issues/78#issuecomment-4259825922), [data analysis](https://github.com/center-for-humans-and-machines/algorithmic-institutions/issues/78#issuecomment-4260030974)

## Switch predictor

TODO — not yet investigated.

## Acceptance criteria

- Contribution simulation reproduces the direction of the pilot trajectory (non-flat, positive slope where pilot shows one)
- Switch predictor positive-class recall ≥ 0.5 AND simulated per-round switch counts within ±30% of pilot at decision rounds
- Regenerate `comparison_pilot_vs_sim.jpg` and `switch_count_per_round.jpg` with improved models as evidence

## Dependencies

- Depends on: #77 (merge first so corrected-dataset baseline is in)
- Blocks: #45 (simulation validation not trustworthy until these gaps close)

🤖🧑

**Comment by cemrtkn** [comment]
## Contribution model grid search & training dynamics comparison

### Grid search results

Ran two rounds of grid search on the group switching contribution model:
1. **Architecture × hidden_size** (12 jobs, 1000 epochs): node+edge+rnn with hidden_size=20 wins clearly (test loss 2.31 vs 2.79 for baseline hidden_size=5)
2. **LR × weight_decay** for top architectures (24 jobs, 5000 epochs): lr=1e-4 dominates across the board; weight_decay has negligible effect

Best config: **node+edge+rnn, hidden_size=20, lr=1e-4, wd=1e-5**

### Training dynamics comparison

Compared three 10k-epoch runs to isolate whether the problem is the data or the hyperparameters:

#### 1. Legacy data + legacy params — healthy convergence
![legacy_contribution_v4](https://github.com/center-for-humans-and-machines/algorithmic-institutions/blob/78-improve-ah-models-contribution-switch/plots/group_selection/comparison/legacy_contribution_v4_loss_cv.png?raw=true)

#### 2. New data + legacy params — wider gap, test flattening
![group_switching_legacy_params](https://github.com/center-for-humans-and-machines/algorithmic-institutions/blob/78-improve-ah-models-contribution-switch/plots/group_selection/comparison/group_switching_contribution_legacy_params_loss_cv.png?raw=true)

#### 3. New data + optimized params — overfitting past ~5000 epochs
![group_switching_optimized](https://github.com/center-for-humans-and-machines/algorithmic-institutions/blob/78-improve-ah-models-contribution-switch/plots/group_selection/comparison/group_switching_contribution_loss_cv.png?raw=true)

| Setup | Train loss | Test loss | Gap | Dynamics |
|---|---|---|---|---|
| Legacy data + legacy params | ~1.75 | ~1.85 | **~0.10** | Healthy convergence, tight gap |
| New data + legacy params | ~1.82 | ~2.38 | ~0.55 | Wider gap, test flattening |
| New data + optimized params | ~1.52 | ~2.50 | ~1.00 | **Overfitting** past epoch ~5000 |

### Key finding

The legacy model on the original pilot data shows textbook-healthy training dynamics — train and test loss converge tightly with minimal generalization gap. Neither configuration on the group switching data comes close to replicating this, regardless of architecture or hyperparameter tuning. The optimized model (hidden_size=20) fits the training data better but overfits, while the legacy params underfit.

### Next steps

The wall we're hitting is likely **data-driven, not model-driven**. The group switching dataset has ~11 episodes vs the much larger legacy pilot set, and the 8-player/2-group structure adds complexity. Before further model tuning, we need to:

1. **Dive into the dataset preprocessing** — understand how the group switching data is constructed and whether the preprocessing pipeline introduces artifacts or information loss
2. **Analyze the data distribution** — compare class distributions, temporal patterns, and feature statistics between legacy and group switching datasets to pinpoint what makes the new data harder to learn
3. **Quantify the data gap** — determine whether this is purely a sample size issue or if structural differences in the experimental design contribute

This should let us prove (or disprove) the hypothesis that the data quantity and structure are the bottleneck, not the model capacity.

**Comment by cemrtkn** [comment]
## Data analysis: legacy vs group switching

### Dataset comparison

| Metric | Legacy | New |
|---|---|---|
| Episodes | 135 | 13 |
| Total rows | 5,844 | 2,496 |
| Players/episode | 4 | 8 |
| Rounds | 16 | 24 |
| Mean contribution | 13.0 | 9.9 |
| Contribution=20 share | 32% | 11% |
| Contribution entropy | 3.67 | 3.99 |
| Contribution trajectory | 11.3→14.5 | 8.7→10.0 |

The new data has 10x fewer episodes. Each episode produces more rows (8 players × 24 rounds vs 4 × 16), but the model learns episode-level patterns — 13 distinct trajectories is the real bottleneck. With 5-fold CV, each test fold has only ~2-3 episodes. Legacy gets ~27 episodes per test fold.

Contribution distribution is more spread out (higher entropy) and the temporal signal is weaker and noisier.

### Feature importance

**New data (group switching):**
```
                feature  mutual_info  rf_importance     rho  avg_rank
      prev_contribution       0.8115         0.5438  0.7351    1.0000
        prev_punishment       0.0466         0.1415 -0.2432    2.3333
            agent_group       0.0345         0.0360  0.1891    3.3333
           round_number       0.0039         0.2657  0.1093    3.3333
prev_contribution_valid       0.0019         0.0130  0.0306    5.0000
```

**Legacy data:**
```
                feature  mutual_info  rf_importance     rho  avg_rank
      prev_contribution       0.6305         0.6691  0.7333    1.0000
        prev_punishment       0.0801         0.1620 -0.3282    2.0000
           round_number       0.0318         0.1551  0.1671    3.0000
prev_contribution_valid       0.0042         0.0138  0.0899    4.0000
```

Feature correlations are comparable across both datasets — the signal is there. The bottleneck is sample size, not feature quality.

Scripts: `scripts/data_analysis/compare_datasets.py`, `scripts/data_analysis/feature_importance.py`

**Comment by cemrtkn** [comment]
### Data quantity experiment

> [!WARNING]
> **Correction (2026-04-21):** The original augmentation experiment below had a data leak — `augment_episodes.py` ran on the full CSV *before* the CV split, so noisy copies of test episodes ended up in the training fold. The reported 2.21 → 2.09 improvement was largely leakage, not genuine generalization. Scroll to the bottom for the corrected, leak-free result.

**Downsampling legacy data** to 13 episodes (matching group switching) and **augmenting group switching data** 3x via noise injection (std=1.0 on contribution/punishment).

| Setup | Train loss | Test loss | Gap | Best test | Best test epoch |
|---|---|---|---|---|---|
| Legacy (135 ep) | ~1.75 | ~1.85 | ~0.10 | ~1.85 | - |
| Legacy downsampled (13 ep) | 1.91 | 2.06 | 0.16 | 2.06 | ~2150 |
| New data + legacy params | ~1.82 | ~2.38 | ~0.55 | ~2.38 | - |
| New data + optimized params | 1.84 | 2.21 | 0.37 | 2.21 | ~4850 |
| ~~New data + augmented 3x~~ (leaky) | ~~1.89~~ | ~~2.09~~ | ~~0.20~~ | ~~2.09~~ | ~~~5150~~ |

Data quantity is still the most likely culprit: downsampled legacy (13 ep) reproduces the same ~2.0 loss floor as the group switching runs. The augmentation claim below had to be retracted.

---

### Correction: leak-free augmentation (2026-04-21)

Moved augmentation into the training pipeline (new `data_augmentation:` block in the training config, applied per-fold on the train split only, after CV splitting). Re-ran the contribution model with 10x noise augmentation, 1000 epochs, otherwise identical hyperparameters.

| Setup | Train loss (final) | Test loss floor | Gap | Dynamics |
|---|---|---|---|---|
| New data, no augmentation (10k ep) | ~1.52 | **~2.25** @ ep ~4500 | ~1.00 | Overfits past ep ~5000 |
| New data + 10x leak-free noise (1k ep) | ~2.00 | **~2.25** @ ep ~1000 | ~0.25 | Plateau, no overfit |

![contribution augmentation comparison](https://github.com/center-for-humans-and-machines/algorithmic-institutions/blob/78-improve-ah-models-contribution-switch/plots/group_selection/comparison/group_switching_contribution_vs_group_switching_contribution_augmented_loss_cv.png?raw=true)

**Both runs hit the same test loss floor (~2.25).** Augmentation:
- Closes the train/test gap (regularization effect ✓)
- Reaches the floor ~4× faster
- **Does not lower the floor** — generalization is unchanged

So augmentation buys us regularization and convergence speed, but not generalization. Data quantity remains the real bottleneck; noise-based upsampling cannot substitute for additional real episodes.

Relevant switch predictor result: the leak-free run on the same augmentation pipeline collapsed from 0.194 (leaky) → 0.558 (leak-free) test log-loss, same story — the previous gains were artifacts.

**Comment by cemrtkn** [comment]
## Switch predictor grid search results (inconclusive)

Ran two grid searches over the switch predictor hyperparameters on branch `78-improve-ah-models-contribution-switch`. Full write-up in `reports/switch_predictor_grid_search.md`.

**Grid v1** — swept architecture (`mlp`, `mlp+rnn`, `mlp+rnn+edge`), `hidden_size` (5, 10, 20), `lr` (1e-3, 3e-3). 18 configs, 8-fold CV.
- Winner: `mlp+rnn+edge`, h=20, lr=3e-3 → test log-loss 0.529
- `lr=3e-3` dominated; `hidden_size=20` best (the #76 shrink to 5 was too aggressive)
- Feature importance: `round_number` showed ~0 delta but removing it degraded production runs → kept

**Grid v2** — locked architecture/lr, swept `add_global_model`, `hidden_size` (20, 40), `weight_decay`, `l1_entropy`. 16 configs, 8-fold CV.
- `add_global_model=True` ranked top within the grid (0.520 vs 0.529)
- `hidden_size=40` overfits earlier without improving test loss
- `weight_decay` and `l1_entropy` had no measurable effect

**Bottom line:** the hyperparameter optimization surface is essentially flat for this model + dataset combination. All reasonable configs cluster within one CV std (~0.05) of each other. The model extracts only ~0.04 nats of signal above the class-frequency prior, consistent with the small training set (~11 episodes, ~440 switch decisions). Switch recall remains at 20-30%, far below the ≥50% acceptance criterion — this gap is driven by class imbalance and data scarcity, not architecture or hyperparameters.

Locked in `opt.yml` with `mlp+rnn+edge`, `add_global_model=True`, `hidden_size=20`, `lr=3e-3`, 75 epochs. Config: `configs/training/artificial_humans/switch_predictor/opt.yml`.
